1523 wait for processor fix

clemens-msupply opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #1523

I was running into this quite consistency now... I first tried to just retry the `check_{*} ` methods in the tests. This worked, but it required to change the semantics of the check methods, e.g. they needed to return a bool to indicated that they should be retried (unwinding the panic didn't work because a mut reference to the Tester was used).

In this PR I just look at the processor queue and wait till the processor handled everything. I used some code from the processor but the usecase is slightly different for which reason I copied the code rather than refactor the processor to have something like an "is_done()" method.

clemens-msupply commented on server/service/src/processors/test_helpers.rs:
```diff
@@ -1,12 +1,32 @@
 use tokio::task::JoinSet;
 
-const PROCESSOR_DELAY_MILLISECONDS: u64 = 300;
+const PROCESSOR_DELAY_MILLISECONDS: u64 = 50;
```
I wait a shorter time but check more often now...

clemens-msupply commented on server/service/src/processors/transfer/requisition/test.rs:
```diff
@@ -97,18 +120,22 @@ async fn requisition_transfer() {
+                .unwrap();
+            delayed_with_retries!(requisition_processed(&ctx.connection), "wait for processor");
             tester.check_response_requisition_not_created(&ctx.connection);
             delay_for_processor().await;
```
Are these `delay_for_processor()` calls actually needed? An update should work when the previous check succeeded, shouldn't it?

andreievg commented:
There was a use case when we might want to wait for processors (for example at the [end of sync](https://github.com/openmsupply/open-msupply/blob/caa3c882a5f11a047f18de213ffe7fea9aca4bea/server/service/src/sync/synchroniser.rs#L198-L202), especially during initialisation), we may want to only mark sync as finished after processors have done their work. This wasn't really that important, but when I thought about this use case I thought it should be a message checking that processors are not busy yet. Since it's still desired functionality I thought it's ok to spend some time to do an example of how it can be done (bare bone processor implementation in `main_before.rs` and `main.rs` with the change (it's not final, there are a few unwraps etc, but it should server the purpose of demonstrating):

[checkprocessors.zip](https://github.com/openmsupply/open-msupply/files/11436936/checkprocessors.zip)

`git diff --no-index main_before.rs main.rs`

```diff
diff --git a/main_before.rs b/main.rs
index 2c2f5be..d2bba2c 100644
--- a/main_before.rs
+++ b/main.rs
@@ -1,9 +1,12 @@
-use std::time::Duration;
+use std::time::{Duration, SystemTime};
 
 use tokio::{
     sync::mpsc::Sender,
-    sync::mpsc::{self, Receiver},
-    task::JoinHandle,
+    sync::{
+        mpsc::{self, Receiver},
+        watch,
+    },
+    task::{JoinError, JoinHandle},
     time::sleep,
 };
 
@@ -11,11 +14,15 @@ use tokio::{
 struct ProcessorsTrigger {
     requisition_transfer: Sender<()>,
     shipment_transfer: Sender<()>,
+    request_check: Sender<()>,
+    check_response: watch::Receiver<SystemTime>,
 }
 
 struct Processors {
     requisition_transfer: Receiver<()>,
     shipment_transfer: Receiver<()>,
+    request_check: Receiver<()>,
+    check_response: watch::Sender<SystemTime>,
 }
 
 const CHANNEL_BUFFER_SIZE: usize = 30;
@@ -28,14 +35,23 @@ impl Processors {
         let (shipment_transfer_sender, shipment_transfer_receiver) =
             mpsc::channel(CHANNEL_BUFFER_SIZE);
 
+        let (request_check_sender, request_check_receiver) = mpsc::channel(CHANNEL_BUFFER_SIZE);
+
+        // Maybe can use something other then watch, but it seems the most appropriate
+        let (check_response_sender, check_response_receiver) = watch::channel(SystemTime::now());
+
         (
             ProcessorsTrigger {
                 requisition_transfer: requisition_transfer_sender,
                 shipment_transfer: shipment_transfer_sender,
+                request_check: request_check_sender,
+                check_response: check_response_receiver,
             },
             Processors {
                 requisition_transfer: requisition_transfer_receiver,
                 shipment_transfer: shipment_transfer_receiver,
+                request_check: request_check_receiver,
+                check_response: check_response_sender,
             },
         )
     }
@@ -44,6 +60,8 @@ impl Processors {
         let Processors {
             mut requisition_transfer,
             mut shipment_transfer,
+            mut request_check,
+            check_response,
         } = self;
 
         tokio::spawn(async move {
@@ -51,7 +69,7 @@ impl Processors {
                 // See test below for reasoning behind biased, even though there is no foreseen use case where
                 // requisition must be processed before shipment, it easy to reason about future use cases if
                 // order is guaranteed when requisition transfer is triggered before shipment transfer (like it is in synchroniser)
-                let result = tokio::select! {
+                let __result = tokio::select! {
                     biased;
                     Some(_) = requisition_transfer.recv() => {
                        sleep(Duration::from_secs(5)).await;
@@ -61,6 +79,9 @@ impl Processors {
                         sleep(Duration::from_secs(5)).await;
                         println!("free")
                     },
+                    Some(_) = request_check.recv() => {
+                        check_response.send(SystemTime::now()).unwrap();
+                    },
                     // None will be returned by recv if channel is closed, this would only really happen if all receivers were dropped
                     else => break,
                 };
@@ -69,6 +90,16 @@ impl Processors {
     }
 }
 
+impl ProcessorsTrigger {
+    async fn await_done(self) -> Result<(), JoinError> {
+        let mut check_response = self.check_response.clone();
+        let result = tokio::spawn(async move { check_response.changed().await.unwrap() });
+        self.request_check.send(()).await.unwrap();
+        // maybe can just check_response.changed().await here, but I don't know if there is a race condition
+        result.await
+    }
+}
+
 #[tokio::main]
 async fn main() {
     let (processors_trigger, processors) = Processors::init();
@@ -85,6 +116,8 @@ async fn main() {
             .await
             .unwrap();
         println!("sent");
+        processors_trigger_closure.await_done().await.unwrap();
+        println!("doneski");
     });
 
     tokio::select! {
```

andreievg commented on server/service/src/processors/transfer/requisition/test.rs:
```diff
@@ -1,23 +1,46 @@
+    let changelog_repo = ChangelogRepository::new(connection);
+    let key_value_store_repo = KeyValueStoreRepository::new(connection);
+
+    let cursor = key_value_store_repo
```
Sometime full processor operation would jump a few cursor, depending on the sync setup sometimes both stores are on the same site (processor unit test) and sometimes they are on different sites (integration test), when they are on the same site multiple operations will happen in the same processor task, each operation will update cursor (it's done this way to have the same logic for both use cases of sync configuration):

https://github.com/openmsupply/open-msupply/blob/caa3c882a5f11a047f18de213ffe7fea9aca4bea/server/service/src/processors/transfer/requisition/mod.rs#L86-L98

There [is a diagram](https://github.com/openmsupply/open-msupply/blob/develop/server/service/src/processors/transfer/requisition/doc/omSupply_requisition_transfer_same_site.png) for both stores on the same site (here requisition will be created, then linked in the same processor operation, jumping two change logs)

And extract from [this diagram](https://github.com/openmsupply/open-msupply/blob/develop/server/service/src/processors/transfer/doc/omSupply_transfer_processors.png) where stores are on different sites (in this case requisition will only be created on supply store site, jumping one change log, and linked when created requisition arrives at order store):

![Screenshot 2023-05-10 at 12 02 28 PM](https://github.com/openmsupply/open-msupply/assets/26194949/55f4e4a6-261b-4ba3-b58d-2799fc821c0d)

clemens-msupply commented:
Thanks @andreievg, I implemented an "await processor queue" as you suggested. I use a `oneshot::channel` for this though...

I also removed the remaining `delay_for_processor()` calls because they didn't seem to be needed...

clemens-msupply commented on server/service/src/processors/mod.rs:
```diff
@@ -106,11 +120,35 @@ impl ProcessorsTrigger {
+    /// Its guaranteed that all queued processor events that where in the queue before calling
+    /// this method are handled when this method returns.
+    /// However, new events might have been added while this method was running.
+    pub async fn await_events_processed(&self) {
```
Is not `pub(crate)` because its currently only used in tests and would otherwise give a warning...

andreievg commented on server/service/src/processors/mod.rs:
```diff
@@ -60,13 +69,15 @@ impl Processors {
                 // See test below for reasoning behind biased, even though there is no foreseen use case where
                 // requisition must be processed before shipment, it easy to reason about future use cases if
                 // order is guaranteed when requisition transfer is triggered before shipment transfer (like it is in synchroniser)
+                // The biased flag also makes sure that `await_process_queue` is only called after all other channels are empty.
```
Thanks for that comment

andreievg commented on server/service/src/processors/mod.rs:
```diff
@@ -106,11 +120,35 @@ impl ProcessorsTrigger {
+    /// this method are handled when this method returns.
+    /// However, new events might have been added while this method was running.
+    pub async fn await_events_processed(&self) {
+        let (sender, receiver) = oneshot::channel();
```
Ohh cool, so we send the channel a sender to respond with ! That's pretty cool.

andreievg reviewed (approved):
Looks great, thanks
